package-versions/grouping-and-overrides/multi-targeting.md

```diff
@@ -0,0 +1,76 @@
+---
+layout: default
+title: Multi-Targeting
+parent: Grouping and Overrides
+grand_parent: Package Versions
+nav_order: 5
+---
+
+{% raw %}
+<script>(function(t,e,s,n){var o,a,c;t.SMCX=t.SMCX||[],e.getElementById(n)||(o=e.getElementsByTagName(s),a=o[o.length-1],c=e.createElement(s),c.type="text/javascript",c.async=!0,c.id=n,c.src="https://widget.surveymonkey.com/collect/website/js/tRaiETqnLgj758hTBazgdzqSORtCawzIPfKe_2BNdVN8wMe8HNIqE4OpuND_2BN6BzW5.js",a.parentNode.insertBefore(c,a))})(window,document,"script","smcx-sdk");</script>
+{% endraw %}
+
+# Multi-Targeting
+
+As exciting new features become available in the .NET ecosystem, your code may need to adapt as package versions sometimes lack proper support.
+
+The examples on this page demonstrate how existing MSBuild features can be combined with grouping to provide powerful multi-targeting capabilities.
+
+## Examples
+
+#### Directory.Packages.props
+```xml
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+    <ItemGroup>
+        <!-- When CentralManagementGroup is unspecified for a <PackageVersion>, the <PackageVersion> is part of the default global group. -->
+        <PackageVersion Include="foo" Version="1.1.1" />
+        <!-- These three CentralManagementGroup's are used for multi-targeting. -->
+        <PackageVersion Include="foo" Version="2.2.2" CentralManagementGroup="A" />
+        <PackageVersion Include="foo" Version="3.3.3" CentralManagementGroup="Anetcoreapp3.1" />
+        <PackageVersion Include="foo" Version="4.4.4" CentralManagementGroup="Anet472" />
+    </ItemGroup>
+</Project>
+```
+
+In this `Directory.Packages.props` file, there are four `CentralManagementGroup` of `<PackageVersion>` defined.
+
+- The global `CentralManagementGroup` is defined automatically and applies to version 1.1.1 of `foo` where no `CentralManagementGroup` is specified.
+- `"A"`, `"Anetcoreapp3.1"`, `"Anet472"` are three different `CentralManagementGroup`. They are named similarly as a convention, but they remain three independent groups. MSBuild does not understand any relationship between these groups.
+
+### Sample Project 11: Inline Multi-Targeting
+
+#### 11.csproj
+```xml
+<Project>
+  <ItemGroup>
+    <PackageReference Include="foo" CentralManagementGroup="Anetcoreapp3.1" Condition="'$(TargetFramework)' == 'netcoreapp3.1'" />
+    <PackageReference Include="foo" CentralManagementGroup="Anet472" Condition="'$(TargetFramework)' == 'net472'" />
```

Comment:
I would more likely place the target framework conditions on the ItemGroup instead as it scales to cleaner xml. If the IDE was excellent at maintaining these as-is, I _might_ leave them alone. :)

Reply:
Totally agree, the current syntax is very confusing
Here the package is the same and only the version change, this should NOT be in the `csproj` but in the `Directory.Packages.props`

With this example,
* the logic about Version stays in the file about CPVM
* It is grouped per TFM
* It's very similar as how MsBuild works 
* the logic of "only reference this if ..." is kept in `csproj`
* less noisy / repeatition

Directory.Packages.props:
```xml
<Project>
  <ItemGroup Condition=" '$(TargetFramework)' == 'net472' ">
    <PackageVersion Include="foo" Version="1.0.0" />
    <PackageVersion Include="bar" Version="2.0.0" />
  </ItemGroup>

  <ItemGroup Condition=" '$(TargetFramework)' == 'net5.0' ">
    <PackageVersion Include="foo" Version="5.0.0" />
  </ItemGroup>
</Project>
```
foo.csproj:
```xml
<Project>
  <PropertyGroup>
    <TargetFrameworks>net472;net5.0<TargetFrameworks>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="foo" />
  </ItemGroup>

  <ItemGroup Condition=" '$(TargetFramework)' == 'net472' ">
    <PackageReference Include="bar" />
  </ItemGroup>
</Project>
```

Reply:
also, the `TargetFrameworks` is evaluated latter on due to how `PropertyGroup` and `ItemGoup` are evaluated
given the fact that 
* `props` file are supposed to be loaded / imported at the beginning of `csproj` files
* `targets`  file are supposed to be loaded / imported at the end of `csproj` files

If the `Directory.Packages.props` is imported before the `csproj` and your project in the solution have a mix of `TargetFrameworks` and `TargetFramework` you can end up with `TargetFramework` being empty

Current `Directory.Packages.props` works almost well without group for Multi targeting to be honest ;)
There's just this little edge case that could be handled by moving to `targets` :)

Reply:
The example from @tebeco would be great for me as well. It would use existing MSBuild features and sounds like a more natural next step instead of having to use groups. It also naturally fits with how I structure multi-targeting project files right now - e.g. https://github.com/opentracing-contrib/csharp-netcore/blob/ebb38ade5d339c2e19d7c9a75297c82c1d4e6692/src/OpenTracing.Contrib.NetCore/OpenTracing.Contrib.NetCore.csproj